File: src/utils/logger.ts
import pino from "pino";

// In stdio MCP, keep stdout exclusive to JSON-RPC, so logs are fixed to stderr.
// External log forwarding, success access logs, and pricing were removed in spec 021; only error logs are written to stderr.
export const logger = pino({
  level: process.env.LOG_LEVEL ?? "info",
  base: null, // Exclude pid and hostname to keep the JSON simple (pino removes all when set to null).
  timestamp: pino.stdTimeFunctions.epochTime,
}, pino.destination(2)); // fd 2 = stderr

export function logError(
  log_type: "mcp_error",
  tool_name: string,
  model: string,
  duration_ms: number,
  error: unknown,
  thinking_level = ""
): void {
  const error_type = error instanceof Error ? error.constructor.name : "Error";
  const error_message = error instanceof Error ? error.message : String(error);
  const entry = {
    log_type,
    tool_name,
    model,
    status: "error" as const,
    duration_ms: Math.round(duration_ms * 1000) / 1000,
    thinking_level,
    error_type,
    error_message,
  };
  logger.error(entry, `[${tool_name}] ${error_type}: ${error_message}`);
}

File: src/utils/logger.rs
//! Structured logging to **stderr only** (stdout is reserved for the MCP JSON-RPC
//! stream). Replaces the original `pino` setup; only error logs are emitted
//! (success access logs / pricing were removed in spec 021).

use crate::utils::errors::GeminiError;

/// Initialize the global tracing subscriber: JSON formatter, stderr writer, level
/// taken from `LOG_LEVEL` (default `info`). Idempotent-safe: a second call is a no-op
/// because `try_init` returns an error that we deliberately ignore.
pub fn init_tracing() {
    use tracing_subscriber::EnvFilter;

    let level = std::env::var("LOG_LEVEL").unwrap_or_else(|_| "info".to_string());
    let filter = EnvFilter::try_new(&level).unwrap_or_else(|_| EnvFilter::new("info"));

    let _ = tracing_subscriber::fmt()
        .json()
        .with_writer(std::io::stderr)
        .with_env_filter(filter)
        .try_init();
}

/// Emit a structured error log for a failed tool invocation.
pub fn log_error(
    tool_name: &str,
    model: &str,
    duration_ms: f64,
    error: &GeminiError,
    thinking_level: &str,
) {
    let error_type = error.error_type();
    let error_message = error.to_string();
    tracing::error!(
        log_type = "mcp_error",
        tool_name,
        model,
        status = "error",
        duration_ms = (duration_ms * 1000.0).round() / 1000.0,
        thinking_level,
        error_type = %error_type,
        error_message = %error_message,
        "[{tool_name}] {error_type}: {error_message}"
    );
}
